pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,476 @@
+// traces.
+// Note that there could be a race between another caller calling Register and
+// us closing the tracer, however we won't allow any new registrations to come
+// through after this call. Note that we set this to MaxInt32 instead of
+// MaxInt64 to avoid a rare race where another thread is in the middle of
+// `reserveSpace` and calls `Add(1)` which cause overflow.
+func (r *RaftTracer) Close() {
+	r.numRegisteredReplica.Store(math.MaxInt32)
+
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		t.logf(2, "cleanup log index %d during Close", index)
+		r.removeEntry(index)
+		return true
+	})
+}
+
+func peer(p raftpb.PeerID) redact.SafeString {
+	return redact.SafeString(raft.DescribeTarget(p))
+}
+
+// traceIfCovered will log the message if it touches any of the registered trace
+// points. Additionally it returns any saved trace/span IDs for sending to
+// remote nodes. This applies both to messages that the leader sends to
+// followers, and messages replicas send to their local storage.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []kvserverpb.TracedEntry {
+	if len(m.Entries) == 0 {
+		return nil
+	}
+	minEntryIndex := kvpb.RaftIndex(m.Entries[0].Index)
+	maxEntryIndex := kvpb.RaftIndex(m.Entries[len(m.Entries)-1].Index)
+	var tracedEntries []kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		// If the traced index is not in the range of the entries, we can skip
+		// it. We don't need to check each individual entry since they are
+		// contiguous.
+		if t.traced.Index < minEntryIndex || t.traced.Index > maxEntryIndex {
+			return true
+		}
+		tracedEntries = append(tracedEntries, t.traced)
+		// TODO(baptist): Not all the fields are relevant to log for all
+		// message types. Consider cleaning up what is logged.
+		t.logf(4,
+			"%s->%s %v Term:%d Log:%d/%d Entries:[%d-%d]",
+			peer(m.From),
+			peer(m.To),
+			m.Type,
+			m.Term,
+			m.LogTerm,
+			m.Index,
+			minEntryIndex,
+			maxEntryIndex,
+		)
+		return true
+	})
+	return tracedEntries
+}
+
+// traceIfPast will log the message the message is past any registered tracing
+// points. It will additionally unregister traces that are no longer useful.
+// This call is for events that move the needle/watermark forward (e.g. the log
+// storage syncs), but don't have an exact range of entries affected. So, being
+// unable to match these events to entries exactly once, we instead check that
+// the watermark passed the entry. To protect against overly verbose logging, we
+// only allow MsgAppResp to be logged once per peer, and only one
+// MsgStorageAppendResp. When we receive a MsgStorageApplyResp we will log and
+// unregister the tracing.
+func (r *RaftTracer) traceIfPast(m raftpb.Message) {
+	if m.Reject {
+		return
+	}
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		switch m.Type {
+		case raftpb.MsgAppResp:
+			if kvpb.RaftIndex(m.Index) >= index && t.seenMsgAppResp(m.From) {
```

Comment:
nit: the "seen" logic is inverted here. Logically, we want to proceed if we **haven't** seen a `MsgAppResp` yet. Consider inverting the meaning of the returned bool so that this line becomes:

```golang
kvpb.RaftIndex(m.Index) >= index && !t.seenMsgAppResp(m.From)
```

Same for the other "seen" method.

Reply:
Yeah - I realized after looking at this agin that I had inverted the meaning I originally had. Anyway I swapped the true/false and the condition and validated that it is good.